File: 114_Flatten Binary Tree to Linked List.java
// 时间复杂度：O(n)，其中 n 是二叉树的节点个数。
// 空间复杂度：O(n)。递归需要 O(n) 的栈空间。

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode head = null;

    public void flatten(TreeNode root) {
        if (root == null) return;

        this.flatten(root.right);
        this.flatten(root.left);

        root.left = null;
        root.right = this.head;
        this.head = root;
    }
}

File: 114_Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Method 1: reverse-pre: right-left-mid (头插法)
    head = None

    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root: return

        self.flatten(root.right)
        self.flatten(root.left)

        root.left = None
        root.right = self.head
        self.head = root


    # # Method 2: simulate pre-order
    # def flatten(self, root: Optional[TreeNode]) -> None:
    #     """
    #     Do not return anything, modify root in-place instead.
    #     """
    #     # Pre-Order
    #     vHead = cur = TreeNode()

    #     def pre(node: Optional[TreeNode]):
    #         if not node: return

    #         nonlocal cur
    #         cur.right = node
    #         cur = cur.right
    #         tmpl, tmpr = cur.left, cur.right
    #         cur.left = cur.right = None

    #         pre(tmpl)
    #         pre(tmpr)
        
    #     pre(root)